File: 백준/Gold/19236. 청소년 상어/청소년 상어.cc
#include <iostream>
#include <algorithm>
using namespace std;
 
struct FISH{
	int y, x, dir;
	bool isdead;
};
 
 
const int dx[] = { 0,-1,-1,-1,0,1,1,1 };
const int dy[] = { -1,-1,0,1,1,1,0,-1 };
 
void cpmap(int c[][4], int o[4][4])
{
    for(int i = 0; i < 4; i++){
        for(int j = 0; j < 4; j++){
            c[i][j] = o[i][j];
        }
    }
}
void cpfish(FISH c[17], FISH o[17])
{
    for(int i = 0; i < 17; i++){
        c[i] = o[i];
    }
}

int dfs(int origin_map[][4], int y, int x, FISH origin_fish[17])
{
	
	int map[4][4];
	FISH fish[17];
	cpmap(map, origin_map);
	cpfish(fish, origin_fish);

	int eat = map[y][x];
	int dir = fish[map[y][x]].dir;
    
    // 1. shark_eat
	fish[map[y][x]].x = -1;
	fish[map[y][x]].y = -1;
	fish[map[y][x]].isdead = true;
	map[y][x] = 0;
 
    
	// 2. fish_move
	for (int i = 1; i <= 16; ++i) {
		if (fish[i].isdead) continue;
        
        // 방향찾기
		int nx, ny;
		for (int d = 0; d < 8; d++){
            int td = (fish[i].dir + d) % 8;
            int tx = fish[i].x + dx[td];
            int ty = fish[i].y + dy[td];
            if(tx < 0 || tx >= 4 || ty < 0 || ty >= 4) continue;
            if(tx == x && ty == y) continue;
            nx = tx;
            ny = ty;
            fish[i].dir = td;
			break;
		}
		
	    if (map[ny][nx] == 0) {
		    swap(fish[i].x, nx);
		    swap(fish[i].y, ny);
		    swap(map[fish[i].y][fish[i].x], map[ny][nx]);
		}
		else {
			int nxt = map[ny][nx];
			swap(fish[i].x, fish[nxt].x);
			swap(fish[i].y, fish[nxt].y);
            swap(map[fish[i].y][fish[i].x], map[fish[nxt].y][fish[nxt].x]);
		}
	}
 
 
    // 3. shark_move
    int ret = 0;
 
    for(int d = 1; d <= 4; d++){
        int nx = x + (dx[dir] * d);
        int ny = y + (dy[dir] * d);
        if(nx < 0 || nx >= 4 || ny < 0 || ny >= 4) continue;
        if(map[ny][nx] == 0) continue;

  	    ret = max(ret, dfs(map, ny, nx, fish));
    }
    
    return ret + eat;
}
 
int main()
{
	FISH fish[17];
    int map[4][4];
    
    for(int i = 0; i < 4; i++){
        for(int j = 0; j < 4; j++){
            map[i][j] = 0;
            int f_n, f_dir;
            cin >> f_n >> f_dir;
            f_dir--;
            fish[f_n] = FISH{i, j, f_dir, 0};
            map[i][j] = f_n;
        }
    }
    
	cout << dfs(map, 0, 0, fish);
	return 0;
}

File: 백준/Gold/19236. 청소년 상어/청소년 상어.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.util.Scanner;

/**
 * 1. 상어 먹음 2. 물고기 이동 3. 상어이동
 * 
 * 상태 변화 많음. dfs(cmap, cfish, sharkX, sharkY, sum) 복사 map = cmap, fish = cfish
 * 
 * // 상어먹음
 * 
 */
public class Main {

	static int ret = 0;
	static int dx[] = new int[] { -1, -1, 0, 1, 1, 1, 0, -1 };
	static int dy[] = new int[] { 0, -1, -1, -1, 0, 1, 1, 1 };

	static class Fish {
		int x, y, dir;

		Fish(int x, int y, int dir) {
			this.x = x;
			this.y = y;
			this.dir = dir;
		}
	}

	public static void main(String[] args) throws Exception {
		Scanner sc = new Scanner(System.in);
		int map[][] = new int[4][4];
		Fish fish[] = new Fish[17];
		// 입력받기 + 초기화
		for (int i = 0; i < 4; i++) {
			for (int j = 0; j < 4; j++) {
				int n = sc.nextInt();
				int d = sc.nextInt() - 1;
				map[i][j] = n;
				fish[n] = new Fish(i, j, d);
			}
		}

		dfs(map, fish, 0, 0, 0);
		System.out.println(ret);
	}

	static void dfs(int[][] cmap, Fish[] cfish, int sharkX, int sharkY, int sum) {

		// 0. 배열 복사
		int[][] map = new int[4][4];
		Fish[] fish = new Fish[17];

		for (int i = 0; i < 4; i++) {
			for (int j = 0; j < 4; j++)
				map[i][j] = cmap[i][j];
		}

		for (int i = 1; i < 17; i++) {
			fish[i] = new Fish(cfish[i].x, cfish[i].y, cfish[i].dir);
		}

		// 1. 상어 먹음
		int eat = map[sharkX][sharkY];
		int sharkDir = fish[eat].dir;
		fish[eat] = new Fish(-1, -1, -1);
		map[sharkX][sharkY] = -1;
		sum += eat;
		ret = Math.max(ret, sum);

		// 2. 물고기 이동
		for (int i = 1; i <= 16; i++) {
			if (fish[i].x == -1)
				continue;
			int cx = fish[i].x;
			int cy = fish[i].y;
			int cd = fish[i].dir;

			int nx = cx + dx[cd];
			int ny = cy + dy[cd];
			int nd = cd;

			while (nx < 0 || nx >= 4 || ny < 0 || ny >= 4 || (sharkX == nx && sharkY == ny)) {
				nd = (nd + 1) % 8;
				nx = cx + dx[nd];
				ny = cy + dy[nd];
			}

			int target = map[nx][ny];
			map[nx][ny] = i;
			map[cx][cy] = target;
			fish[i] = new Fish(nx, ny, nd);
			if (target != -1) {
				fish[target] = new Fish(cx, cy, fish[target].dir);
			}
		}

		// 3. 상어 이동
		for (int i = 1; i <= 3; i++) {
			int nx = sharkX + dx[sharkDir] * i;
			int ny = sharkY + dy[sharkDir] * i;

			if (nx < 0 || nx >= 4 || ny < 0 || ny >= 4 || map[nx][ny] == -1)
				continue;

			dfs(map, fish, nx, ny, sum);
		}
	}
}